src/java.desktop/share/classes/sun/swing/CachedPainter.java

```diff
@@ -314,8 +314,9 @@ public int getHeight(ImageObserver observer) {
 
         @Override
         public Image getResolutionVariant(double destWidth, double destHeight) {
-            int w = (int) Math.ceil(destWidth);
-            int h = (int) Math.ceil(destHeight);
+            int w = (int) Math.floor(destWidth + 0.5);
+            int h = (int) Math.floor(destHeight + 0.5);
```

Comment:
No, Region.clipRound does something little different than what I intend to do in
this code . 
It works same as my code for 0.25 and 0.75 case. But not for 0.5 case.
For example in case of destwidth and destHeight  = 19.5, 
for my current code I see them rounded to 20, while with clipRound it is 
rounded to 19.

Reply:
I guess it should do the same rounding the component do. If the size of the component is 19.5 and we round it to 20, then the image should use the same round logic. BTW why 19 is better than 20 here?

Reply:
So, I added this code specifically to have improved rendering for the 25% cases, like 125%, 225%, 325% as rendering was a little off for these scale values and empirically based on what we saw it looked better with this calculation.
It is also important to note that windows is not consistent in what size it provides for a part(button) . For example for 150% it provides part size as 20 and we calculate it as 20 in above code.  For 175% windows still provides 20 again (while we would expect higher value) and we calculate it as 23 and it works out good. But as I said earlier for 125% or 225% it works best with above code.
The 20 vs 19 thing that I mentioned above was to just show that clipRound doesn't calculate it same was as my code and can have rendering repercussions as it is very sensitive to pixel variation , even 1 pixel variation causes  rendering artifacts.

Reply:
>So, I added this code specifically to have improved rendering for the 25% cases, like 125%, 225%, 325% as rendering was a little off f

But why it looks off if the clipRound is used? Does that size scales worse or does it cut the image at some direction?

Reply:
From the picture above it does not seem that the difference is just 1 pixel, no?

Reply:
Perhaps, this needs additional details. The size of checkbox and radio button is 13 as reported by Theme API for the standard DPI of 96. Being a prime number, 13 doesn't scale well, there's always fractional part unless the scale is whole. Thus, the scaled size never matches.

As the result, `CachedPainter` seems to apply a scale to compensate for that — whether we choose ceil or floot.

The entire thing should have been simpler: a `MultiResolutionImage` which paints the background according to the scale requested. Yet it proved to be not as simple. We're planning to look further into simplifying it if possible and, alternatively or in addition to, considering [dynamic layout based on graphics configuration](https://github.com/openjdk/jdk/pull/13701#discussion_r1187985627) as you suggested above.

The current solution may be not perfect but it provides better rendering compared to what we have now: *the controls look crisp only on the main display*.

Reply:
But still why the size on the components are so different? is it really caused by one "pixel/unit" we requested from the native?

Reply:
So, basically for w,h =19 we get a DPI of 140 and Dimension d = getPartSize(getTheme(widget, dpi), part, state); at line 172 of ThemeReader.java call returns a part size of 13 x 13 provided by windows as it is closer to DPI=120 (for 125% scaling).
While for  w,h =20 we get a DPI of 147 and Dimension d = getPartSize(getTheme(widget, dpi), part, state); at line 172 of ThemeReader.java call returns a part size of 20 x 20 provided by windows as it is closer to DPI=144 (for 150% scaling).
Thus part size calculated by us and provided by Windows matches and hence we see bigger properly rendered buttons while for the w,h=19 case we have 19x19 , while windows gives 13x13 
and hence you see scaled down buttons with bad rendering.

Reply:
Probably I missed something but I would like to clarify. As you mention above we ask Windows to return the image for the component for some specific size, the Windows may return image of exactly requested size if available or any other size. What we will do if the size is different? In the image above it seems we use it as is, but should we? probably we should rescale it?

Reply:
So, we ask Windows for the component image based on a particular DPI , not some specific size.
The basic premise here is that we expect Windows returning image that would match the DPI , and then we perform our calculations accordingly. In the case above (if Region.clipRound is used) that basic premise is not met (if w, h=19) due to us calculating a DPI value of 140 which is closer to 120 or 125% scaling while our scaling is acytually 150% and Windows returns us a smaller size of the component than expected.
So we don't need to add extra code to rescale , I believe.
If we provide windows with correctly calculated DPI value which my code change does it should work fine as Windows will give us image of the correct size.

Reply:
The above code calls getImage() and passes width and height to it.

This is the flow :
getImage(){CachedPainter.java} -> paintToImage() {XPStyle.java} -> ThemeReader.paintBackground ()

We calculate and set scale in getImage() with this:

Line 172, CachedPainter.java

 if (volatileImage == null) {
     if ((w != baseWidth || h != baseHeight)) {
         g2.scale((double) w / baseWidth, 
                 (double) h / baseHeight); 
     }
     paintToImage(c, image, g2, baseWidth, baseHeight, args);
 }
 
 So, scale changes based on height and width of the image which we get from the
 code you mentioned above.
 
 This scale is then used in paintToImage() to calculate DPI at Line 711 of XPStyle.java:
 
   // Get DPI to pass further to ThemeReader.paintBackground()
  Graphics2D g2d = (Graphics2D) g;
  AffineTransform at = g2d.getTransform();
  int dpi = (int)(at.getScaleX() * 96);
 
 
 This DPI is then used to get the part size from windows and it changes as I explained in the previous comment.

Reply:
II am not sure that is correct way to calculate the dpi based on the "adhoc" change of the image size. That will add the mismatch between the code where we create the image, and the code where we request the image. We probably should change them both, or calculate the dpi in some other way.

Also the fact that on the image above one component is rendered twice smaller that another make me think that we did not rescale the image if the Windows return the image of the different size(You probably can prove that by always returning the wrong size from the native and check how it will look on HiDPI screen)

Reply:
> II am not sure that is correct way to calculate the dpi based on the "adhoc" change of the image size. That will add the mismatch between the code where we create the image, and the code where we request the image.

There's always a mismatch (at this time). We already discussed it [above](https://github.com/openjdk/jdk/pull/13701#discussion_r1195604571).

For example, for 125% scale, Windows returns 13×13 like for 100%; yet Swing expects 16.25. This is handled by rendering the 13×13-sized theme background into 16×16-sized bitmap — *to avoid any scale transformations being applied* to keep the image *crisp*.

> We probably should change them both, or calculate the dpi in some other way.

The DPI is calculated based on the scale set on the Graphics. Rajat provided the relevant excerpt:

https://github.com/openjdk/jdk/blob/727836ea85394b8baacd19a834f2993f06084c44/src/java.desktop/windows/classes/com/sun/java/swing/plaf/windows/XPStyle.java#L708-L711

This scale is calculated based on the size of the image.

So, we probably should… change. As [I already said](https://github.com/openjdk/jdk/pull/13701#discussion_r1195604571), this fix should've been as simple as a `MultiResolutionImage` where [`getResolutionVariant`](https://github.com/openjdk/jdk/blob/a780bf0dda4c2c3dce4337ccea37bcdce52045de/src/java.desktop/share/classes/sun/swing/CachedPainter.java#L316) returns the image of the correct size if it's already stored or retrieves it from the system. But it's not the case, and `CachedPainter` seems to be a hindrance rather than a helper.

That being said, we're planning to look further into simplifying it. Perhaps, `XPStyle.SkinPainter` shouldn't extend `CachedPainter`…

> Also the fact that on the image above one component is rendered twice smaller that another make me think that we did not rescale the image if the Windows return the image of the different size(You probably can prove that by always returning the wrong size from the native and check how it will look on HiDPI screen)

The point of the this fix is to *avoid rescaling* what Windows returns. And it *succeeds*: check boxes and radio buttons look much better — sharper and crisper — when you drag Swing UI frame from a display with one DPI to that with another; they even look better on the main display if its DPI is 120 or 168, that is 125% and 175% correspondingly.

Reply:
> > II am not sure that is correct way to calculate the dpi based on the "adhoc" change of the image size. That will add the mismatch between the code where we create the image, and the code where we request the image.
> 
> There's always a mismatch (at this time). We already discussed it [above](https://github.com/openjdk/jdk/pull/13701#discussion_r1195604571).
> 
> For example, for 125% scale, Windows returns 13×13 like for 100%; yet Swing expects 16.25. This is handled by rendering the 13×13-sized theme background into 16×16-sized bitmap — _to avoid any scale transformations being applied_ to keep the image _crisp_.

Are you sure? Currently the code for image [creation](https://github.com/openjdk/jdk/blob/79a4ac791c826656b3e984fe54dc472c62efd028/src/java.desktop/share/classes/sun/swing/CachedPainter.java#L155) and the image [requests](https://github.com/openjdk/jdk/blob/79a4ac791c826656b3e984fe54dc472c62efd028/src/java.desktop/share/classes/sun/swing/CachedPainter.java#L317) use the same "ceil" logic.

> The DPI is calculated based on the scale set on the Graphics. Rajat provided the relevant excerpt:
> 
> https://github.com/openjdk/jdk/blob/727836ea85394b8baacd19a834f2993f06084c44/src/java.desktop/windows/classes/com/sun/java/swing/plaf/windows/XPStyle.java#L708-L711
> 
> This scale is calculated based on the size of the image.

I am still would like to get clarification about the size we changed by the patch, is it a size of the destination area we should render the component to, or something different. If it is a size of the destination area how we can change that to calculate the DPI? 

> > Also the fact that on the image above one component is rendered twice smaller that another make me think that we did not rescale the image if the Windows return the image of the different size(You probably can prove that by always returning the wrong size from the native and check how it will look on HiDPI screen)
> 
> The point of the this fix is to _avoid rescaling_ what Windows returns. And it _succeeds_: check boxes and radio buttons look much better — sharper and crisper — when you drag Swing UI frame from a display with one DPI to that with another; they even look better on the main display if its DPI is 120 or 168, that is 125% and 175% correspondingly.

Since we cannot guarantee that the Windows always return exact size we should have a backup code to rescale the image.

Reply:
> > There's always a mismatch (at this time). We already discussed it [above](https://github.com/openjdk/jdk/pull/13701#discussion_r1195604571).
> >
> > For example, for 125% scale, Windows returns 13×13 like for 100%; yet Swing expects 16.25. This is handled by rendering the 13×13-sized theme background into 16×16-sized bitmap — _to avoid any scale transformations being applied_ to keep the image _crisp_.
> 
> Are you sure? Currently the code for image [creation](https://github.com/openjdk/jdk/blob/79a4ac791c826656b3e984fe54dc472c62efd028/src/java.desktop/share/classes/sun/swing/CachedPainter.java#L155) and the image [requests](https://github.com/openjdk/jdk/blob/79a4ac791c826656b3e984fe54dc472c62efd028/src/java.desktop/share/classes/sun/swing/CachedPainter.java#L317) use the same "ceil" logic.

Yes, I am. The base width and height are 13×13, in 150% the parameters to `getResolutionVariant` are 16.25 and 16.25 — so the image created was 17×17.

Yet, as you can see, it doesn't match the destination width and height, so a scale to compensate for it is applied in `CachedPainter`.

With the updated formula, the image create is 16×16 which is closer to the requested size. The scale to compensate the different is still applied but it results in less distortion, which eventually looks better.

> > > Also the fact that on the image above one component is rendered twice smaller that another make me think that we did not rescale the image if the Windows return the image of the different size(You probably can prove that by always returning the wrong size from the native and check how it will look on HiDPI screen)
> > 
> > The point of the this fix is to _avoid rescaling_ what Windows returns. And it _succeeds_: check boxes and radio buttons look much better — sharper and crisper — when you drag Swing UI frame from a display with one DPI to that with another; they even look better on the main display if its DPI is 120 or 168, that is 125% and 175% correspondingly.
> 
> Since we cannot guarantee that the Windows always return exact size we should have a backup code to rescale the image.

No, we should not have a backup code to rescale the image: we request Windows to paint the image in the resolution appropriate for the DPI set for a display we're rendering for. If, for whatever reason, Windows returns an image that is too small for this DPI, you will notice it in all Windows native controls and apps, Java apps will not be different.